### PR TITLE
Define and validate security realm

### PR DESCRIPTION
**Pull request by kunsingh**
https://github.com/canoo/dolphin-platform/issues/787

---
This change is [Reviewable](https://reviewable.io/reviews/canoo/dolphin-platform/834)

### CODE REVIEW

File: platform-extras/dolphin-platform-security-client/src/main/java/com/canoo/platform/client/security/Security.java [hunk 1]
```diff
@@ -28,6 +28,8 @@
 
     boolean isAuthorized();
 
+    String getRealmName();
```
Comment (hendrikebbers) [inline platform-extras/dolphin-platform-security-client/src/main/java/com/canoo/platform/client/security/Security.java]: this should be directly annotated with @Deprecated

File: platform-extras/dolphin-platform-security-server/src/main/java/com/canoo/dp/impl/server/security/DolphinKeycloakConfigResolver.java [hunk 2]
```diff
@@ -51,14 +51,23 @@ public KeycloakDeployment resolve(final HttpFacade.Request request) {
         Optional.ofNullable(authEndPoint).orElseThrow(() -> new SecurityException("Auth endpoint for security check is not configured!"));
 
         final AdapterConfig adapterConfig = new AdapterConfig();
-        adapterConfig.setRealm(realmName);
+        if(isRealmExist(realmName)){
+            adapterConfig.setRealm(realmName);
+        }else{
+            throw new SecurityException("Access Denied");
```
Comment (hendrikebbers) [inline platform-extras/dolphin-platform-security-server/src/main/java/com/canoo/dp/impl/server/security/DolphinKeycloakConfigResolver.java]: Here we can add some more info like "Access Denied! The given realm is not in the allowed realms"

File: platform-extras/dolphin-platform-security-server/src/main/java/com/canoo/dp/impl/server/security/DolphinKeycloakConfigResolver.java [hunk 3]
```diff
@@ -51,14 +51,23 @@ public KeycloakDeployment resolve(final HttpFacade.Request request) {
...
+        }
+
         adapterConfig.setResource(applicationName);
         adapterConfig.setAuthServerUrl(authEndPoint);
         Optional.ofNullable(request.getHeader(BEARER_ONLY_HEADER)).
                 ifPresent(v -> adapterConfig.setBearerOnly(true));
         return KeycloakDeploymentBuilder.build(adapterConfig);
     }
 
+    public static boolean isRealmExist(final String realmName){
```
Comment (hendrikebbers) [inline platform-extras/dolphin-platform-security-server/src/main/java/com/canoo/dp/impl/server/security/DolphinKeycloakConfigResolver.java]: isRealmExist > isRealmAllowed

File: platform-extras/dolphin-platform-security-server/src/main/java/com/canoo/dp/impl/server/security/KeycloakConfiguration.java [hunk 4]
```diff
@@ -67,6 +69,10 @@
     public KeycloakConfiguration(final PlatformConfiguration platformConfiguration) {
         Assert.requireNonNull(platformConfiguration, "platformConfiguration");
         this.realmName = platformConfiguration.getProperty(REALM_PROPERTY_NAME, REALM_PROPERTY_DEFAULT_VALUE);
+        this.realmNames = platformConfiguration.getListProperty(REALM_PROPERTY_NAME, Collections.emptyList());
+        if(this.realmNames.isEmpty()){
+            this.realmNames.add(this.realmName);
+        }
```
Comment (hendrikebbers) [inline platform-extras/dolphin-platform-security-server/src/main/java/com/canoo/dp/impl/server/security/KeycloakConfiguration.java]: wrong.
The name should be always added when it is defined:
```
if(this.realmName != null && !this.realmName.isEmpty()){
             this.realmNames.add(this.realmName);
 }
```

File: platform-extras/dolphin-platform-security-server/src/main/java/com/canoo/dp/impl/server/security/KeycloakConfiguration.java [hunk 5]
```diff
@@ -67,6 +69,10 @@
     public KeycloakConfiguration(final PlatformConfiguration platformConfiguration) {
         Assert.requireNonNull(platformConfiguration, "platformConfiguration");
         this.realmName = platformConfiguration.getProperty(REALM_PROPERTY_NAME, REALM_PROPERTY_DEFAULT_VALUE);
```
Comment (hendrikebbers) [inline platform-extras/dolphin-platform-security-server/src/main/java/com/canoo/dp/impl/server/security/KeycloakConfiguration.java]: We should remove the default value. Now it should be possible to not define a realm

File: platform-extras/dolphin-platform-security-client/src/main/java/com/canoo/platform/client/security/Security.java (hunk 1 above)
Comment (hendrikebbers) [reply platform-extras/dolphin-platform-security-client/src/main/java/com/canoo/platform/client/security/Security.java]: In addition: where is this used?

File: platform-extras/dolphin-platform-security-server/src/main/java/com/canoo/dp/impl/server/security/DolphinKeycloakConfigResolver.java (hunk 2 above)
Comment (kunsingh) [reply platform-extras/dolphin-platform-security-server/src/main/java/com/canoo/dp/impl/server/security/DolphinKeycloakConfigResolver.java]: Done.

File: platform-extras/dolphin-platform-security-server/src/main/java/com/canoo/dp/impl/server/security/DolphinKeycloakConfigResolver.java (hunk 3 above)
Comment (kunsingh) [reply platform-extras/dolphin-platform-security-server/src/main/java/com/canoo/dp/impl/server/security/DolphinKeycloakConfigResolver.java]: Done.

File: platform-extras/dolphin-platform-security-server/src/main/java/com/canoo/dp/impl/server/security/KeycloakConfiguration.java (hunk 5 above)
Comment (kunsingh) [reply platform-extras/dolphin-platform-security-server/src/main/java/com/canoo/dp/impl/server/security/KeycloakConfiguration.java]: Done.

File: platform-extras/dolphin-platform-security-server/src/main/java/com/canoo/dp/impl/server/security/KeycloakConfiguration.java (hunk 4 above)
Comment (kunsingh) [reply platform-extras/dolphin-platform-security-server/src/main/java/com/canoo/dp/impl/server/security/KeycloakConfiguration.java]: Done.

File: platform-extras/dolphin-platform-security-client/src/main/java/com/canoo/platform/client/security/Security.java (hunk 1 above)
Comment (kunsingh) [reply platform-extras/dolphin-platform-security-client/src/main/java/com/canoo/platform/client/security/Security.java]: @deprecated is added to security interface.
Currently I have just implemented to KeycloakSecurity like below
```
    @Override
    public String getRealmName() {
        return realmName;
    }
```

File: platform-extras/dolphin-platform-security-client/src/main/java/com/canoo/platform/client/security/Security.java [hunk 6]
```diff
@@ -28,6 +28,9 @@
 
     boolean isAuthorized();
 
+    @Deprecated
+    String getRealmName();
```
Comment (hendrikebbers) [inline platform-extras/dolphin-platform-security-client/src/main/java/com/canoo/platform/client/security/Security.java]: I can not see that this method is used somewhere? Why is this added?

Comment (kunsingh) [reply platform-extras/dolphin-platform-security-client/src/main/java/com/canoo/platform/client/security/Security.java]: ohh..:)  you remember in our discussion you have mentioned that we need something on client side to return realm name. May be I misunderstood something. Shall I remove this?

Comment (hendrikebbers) [reply platform-extras/dolphin-platform-security-client/src/main/java/com/canoo/platform/client/security/Security.java]: When it is not used until now it should be removed :)

Comment (kunsingh) [reply platform-extras/dolphin-platform-security-client/src/main/java/com/canoo/platform/client/security/Security.java]: Done.

**Pull request merged**